### PR TITLE
Sample library project

### PR DESCRIPTION
**Pull request by serhii-lekariev**
This PR defines a sample project that describes a library.

The library is able to:

1) Store books
2) Allow users to borrow the books, controlling for late returners.
3) Allow users to return the books back to the library.

This project showcases the use of a process manager to describe a process of borrowing a book, starting from a `BorrowBook` command and ending with a `ReturnBook` command. The entire process is managed by the respective manager.

Borrowing a book is a process that consists of borrowing a book and returning a book.

A borrowing half of the borrowing process
<p align="center"> Borrowing a book </p>

A returning half of the borrowing process
<p align="center">Returning a borrowed book</p>

### CODE REVIEW

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL.

**Comment by dmdashenkov** [comment]
@serhii-lekariev, please assign yourself to this PR.

**Comment by dmdashenkov** [comment]
@serhii-lekariev, please add the graphic representation of the message flow to the PR description.

File: library/src/main/java/io/spine/examples/library/BookBorrowingProcman.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            for (BookOnHold book : books.getBookOnHoldList()) {
+                ComparableDate today = ComparableDate.today();
+                ComparableDate dateOfBorrowing = new ComparableDate(book.getDateOfBorrowing());
+                if (today.minus(dateOfBorrowing) > DAYS_LIMIT) {
+                    overdue++;
+                }
+            }
+            overdue += books.getOverdueBookCount();
+        }
+        return overdue < 3;
```
Comment (dmdashenkov) [inline library/src/main/java/io/spine/examples/library/BookBorrowingProcman.java]: Please extract the magic number.

File: library/src/main/java/io/spine/examples/library/BookBorrowingRepository.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import io.spine.examples.library.events.BookFetched;
+import io.spine.examples.library.events.PatronDisallowedToBorrow;
+import io.spine.server.procman.ProcessManagerRepository;
+
+/**
+ * A repository that manages {@link io.spine.examples.library.BookBorrowingProcman book borrowing
+ * process managers}.
+ */
+public class BookBorrowingRepository extends ProcessManagerRepository<BookBorrowingId,
+        BookBorrowingProcman,
```
Comment (dmdashenkov) [inline library/src/main/java/io/spine/examples/library/BookBorrowingRepository.java]: Please align the type parameters under one another.

File: library/src/main/java/io/spine/examples/library/BookBorrowingRepository.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                                (message, context) -> ImmutableSet.of(message.getBorrowingId()));
+    }
+
+    @Override
+    protected BookBorrowingProcman findOrCreate(BookBorrowingId id) {
+        BookBorrowingProcman result = super.findOrCreate(id);
+        result.setPatronBorrowings(patronBorrowings);
+        return result;
+    }
+
```
Comment (dmdashenkov) [inline library/src/main/java/io/spine/examples/library/BookBorrowingRepository.java]: Remove empty line.

File: library/src/main/java/io/spine/examples/library/BookRepository.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+package io.spine.examples.library;
+
+import com.google.common.collect.ImmutableSet;
+import io.spine.examples.library.commands.FetchBook;
+import io.spine.examples.library.events.BookReturnedInTime;
+import io.spine.examples.library.events.BookReturnedLate;
+import io.spine.server.aggregate.AggregateRepository;
+
+/**
+ * A repository that manages {@link io.spine.examples.library.BookAggregate book aggregates}.
```
Comment (dmdashenkov) [inline library/src/main/java/io/spine/examples/library/BookRepository.java]: Use `{@linkplain ...}` for such purposes.

File: library/src/main/java/io/spine/examples/library/BookRepository.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+class BookRepository extends AggregateRepository<BookId, BookAggregate> {
+
+    @Override
+    public void onRegistered() {
+        super.onRegistered();
+        getCommandRouting().route(FetchBook.class, (message, context) -> message.getBookId());
+        getEventRouting().route(BookReturnedInTime.class, (message, context) -> ImmutableSet.of(message.getBookId()));
+        getEventRouting().route(BookReturnedLate.class, (message, context) -> ImmutableSet.of(message.getBookId()));
```
Comment (dmdashenkov) [inline library/src/main/java/io/spine/examples/library/BookRepository.java]: These lines are too long.

File: library/src/main/java/io/spine/examples/library/package-info.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
```
Comment (dmdashenkov) [inline library/src/main/java/io/spine/examples/library/package-info.java]: There must be an empty line between the copyright and the doc.

File: library/src/main/java/io/spine/examples/library/package-info.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * This package contains classes that define a library application, that allows to store,
+ * borrow and return borrowed books.
+ */
+package io.spine.examples.library;
```
Comment (dmdashenkov) [inline library/src/main/java/io/spine/examples/library/package-info.java]: Please see the examples of a proper `package-info` in Spine. There are a few annotations to add.

File: library/src/main/proto/spine_examples/library/book.proto [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+option java_multiple_files = true;
+
+message BookId {
+    string value = 1 [(required) = true];
+}
+
+// A book entity.
+//
+message Book {
+
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/book.proto]: Redundant empty line.

File: library/src/main/proto/spine_examples/library/book.proto [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+option (type_url_prefix) = "type.spine.io";
+option java_package = "io.spine.examples.library";
+option java_outer_classname = "BookProto";
+option java_multiple_files = true;
+
+message BookId {
+    string value = 1 [(required) = true];
+}
+
+// A book entity.
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/book.proto]: This doc says nothing. Please describe a `Book` in the real world avoiding the word `entity`.

File: library/src/main/proto/spine_examples/library/book.proto [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    option (entity).kind = AGGREGATE;
+
+    // An identifier of the book.
+    BookId book_id = 1 [(required) = true];
+
+    // Title of the book.
+    string title = 2 [(required) = true];
+
+    // Name of the author who wrote the book.
+    string author = 3 [(required) = true];
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/book.proto]: We have `PersonName` for that. Also, some books might not have an author and some might have several.

File: library/src/main/proto/spine_examples/library/book.proto [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+//
+message Book {
+
+    option (entity).kind = AGGREGATE;
+
+    // An identifier of the book.
+    BookId book_id = 1 [(required) = true];
+
+    // Title of the book.
+    string title = 2 [(required) = true];
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/book.proto]: Is it `title` or `name`? In any case, please use a tiny type.

File: library/src/main/proto/spine_examples/library/book_borrowing.proto [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    // User who borrowed the book.
+    LibraryPatronId book_borrower = 3 [(required) = true];
+
+    // A date the book was borrowed on.
+    spine.time.LocalDate date_of_borrowing = 4 [(required) = true];
+
+    // A library where the borrowed book is stored.
+    LibraryId library_id = 5 [(required) = true];
+
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/book_borrowing.proto]: Please remove redundant empty lines.

File: library/src/main/proto/spine_examples/library/book_borrowing.proto [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+import "spine_examples/library/book.proto";
+import "spine_examples/library/library_patron.proto";
+import "spine_examples/library/library.proto";
+import "spine/time/time.proto";
+
+message BookBorrowingId {
+    string value = 1 [(required) = true];
+}
+
+// A borrowing of some book.
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/book_borrowing.proto]:
```suggestion
// The process of borrowing a book.
```

File: library/src/main/proto/spine_examples/library/commands.proto [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+syntax = "proto3";
+
+package spine_examples.library;
+
+import "spine/options.proto";
+
+option (type_url_prefix) = "type.spine.io";
+option java_package = "io.spine.examples.library.commands";
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/commands.proto]:
```suggestion
option java_package = "io.spine.examples.library.command";
```

Please see [this guide](https://spine.io/docs/guides/naming-conventions.html#Javapackages) for more.

File: library/src/main/proto/spine_examples/library/commands.proto [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+// A command to create a new library.
+message CreateLibrary {
+
+    // An ID of a new library.
+    LibraryId library_id = 1 [(required) = true];
+}
+
+// A command to publish a new book.
+// Unpublished books are considered non-existent.
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/commands.proto]: There must be an empty comment line between the first and the second paragraphs.

File: library/src/main/proto/spine_examples/library/enrichments.proto [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    BookDetails book_details = 1 [(by) = "spine_examples.library.BookAdded.book_id"];
+}
+
+message BookDetails {
+    string author = 1;
+
+    string title = 2;
+}
+
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/enrichments.proto]: Please fix formatting in this file.

File: library/src/main/proto/spine_examples/library/patron_borrowings.proto [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+option java_outer_classname = "BooksOnHoldProto";
+option java_multiple_files = true;
+
+import "spine_examples/library/book.proto";
+import "spine_examples/library/library_patron.proto";
+import "spine/time/time.proto";
+
+message BookOnHold {
+    BookId book_id = 1 [(required) = true];
+    spine.time.LocalDate date_of_borrowing = 2 [(required) = true];
```
Comment (dmdashenkov) [inline library/src/main/proto/spine_examples/library/patron_borrowings.proto]: This is not formatted properly.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL.

File: library/src/test/java/io/spine/examples/library/given/Borrowings.java [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+import io.spine.examples.library.LibraryId;
+import io.spine.examples.library.LibraryPatronId;
+import io.spine.examples.library.command.BorrowBook;
+import io.spine.examples.library.command.ReturnBook;
+import javafx.util.Pair;
+
+import java.util.List;
+import java.util.UUID;
+
+public class Borrowings {
```
Comment (armiol) [inline library/src/test/java/io/spine/examples/library/given/Borrowings.java]: Looks `final`.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL.

**Comment by serhii-lekariev** [comment]
@armiol, PTAL.

**Comment by serhii-lekariev** [comment]
This PR was meant to illustrate a design flow typical for Spine-based projects - starting from designing a model that represents the domain with a desired degree of accuracy, up until testing the whole thing with a `BlackBoxBoundedContext`.

This has been done and PR is therefore closed.

**Pull request closed without merging**